encoder/vaapiencoder_vp9.cpp

```diff
@@ -191,23 +192,56 @@ bool VaapiEncoderVP9::fill(VAEncPictureParameterBufferVP9* picParam,
             picParam->reference_frames[i] = VA_INVALID_SURFACE;
     }
     else {
-        picParam->refresh_frame_flags = 0x01; // refresh last frame
+
         picParam->pic_flags.bits.frame_type = kInterFrame;
 
         ReferenceQueue::const_iterator it = m_reference.begin();
 
-        for (uint32_t i = 0; it != m_reference.end(); ++it, i++)
+        for (uint32_t i = 0; it != m_reference.end(); ++it, i++) {
             picParam->reference_frames[i] = (*it)->getID();
+#if __ENABLE_DEBUG__
+            DEBUG("reference frame[%d] 0x%x", i, (*it)->getID());
+#endif
+        }
 
+        picParam->pic_flags.bits.frame_context_idx = 0;
         // last/golden/alt is used as reference frame. L0 forward
         picParam->ref_flags.bits.ref_frame_ctrl_l0 = 0x7;
 
         // golden and alt are last KeyFrame
         // last is last decoded frame
-        picParam->ref_flags.bits.ref_last_idx = 0;
-        picParam->ref_flags.bits.ref_gf_idx = 1;
-        picParam->ref_flags.bits.ref_arf_idx = 2;
-        picParam->pic_flags.bits.frame_context_idx = 0;
+
+        if (getReferenceMode()) {
```

Comment:
You can use a member variable to keep the refresh_frame_flags. the code should like this in referenceListUpdate
```
if (pic->m_type == VAAPI_PICTURE_I) {
          m_reference.clear();
          m_reference.insert(m_reference.end(), kMaxReferenceFrames, recon);
          m_refreshFrameFlags = (1<< kMaxReferenceFrames) – 1;	
          resetIndexArray();
} else {
         updateIndexArray();
         m_reference[m_index[0]] = recon;
         m_refreshFrameFlags = 1 << m_index[0];
}

```resetIndexArray will set index array to 0, 1, 2
updateIndexArray will do nothing for mode 0 and pop back and push front for mode 1. Only place need care about reference mode  in whole file is updateIndexArray.